serviceregistration/service_registration.go

```diff
@@ -1,40 +1,94 @@
 package serviceregistration
 
+/*
+ServiceRegistration is an interface that can be fulfilled to use
+varying applications for service discovery, regardless of the physical
+back-end used. It uses [the observer pattern](https://refactoring.guru/design-patterns/observer).
+
+Implementing the Factory and adding your factory to the list in commands.go
+is essentially how you register your implementation as an observer. There is
+no deregistration because service discovery stops on its own if Vault stops.
+
+Service registration implements notifications for changes in _dynamic_
+properties regarding Vault's health. Vault's version is the only static
+property given in state for now, but if there's a need for more in the future,
+we could add them on.
+*/
+
 import (
 	"sync"
 
 	log "github.com/hashicorp/go-hclog"
 )
 
+type State struct {
+	VaultVersion                                            string
+	IsInitialized, IsSealed, IsActive, IsPerformanceStandby bool
+}
+
 // Factory is the factory function to create a ServiceRegistration.
-type Factory func(config map[string]string, logger log.Logger) (ServiceRegistration, error)
+//
+// The shutdownCh is the channel to watch for graceful shutdowns _of Vault_,
+// and is great to use for creating background cleanup processes, or for stopping
+// any ongoing goroutines.
+//
+// The config is the key/value pairs set _inside_ the service registration config stanza.
+//
+// The state is only the initial state. The pointer won't be updated over time. All
```

Comment:
Consider changing Factory to accept `State` instead of `*State`, and then it will be implicit in the call that it won't/can't be updated.

Reply:
Oh, good idea, will do.

Reply:
Implemented in https://github.com/hashicorp/vault/pull/8123/commits/2b0f648391e270e6c60ecfdbcd67f76c75170133